File: CALQULUS-PMS/src/mobile/apple/AuthManager.swift
import SwiftUI
import LocalAuthentication
import Security

class AuthManager: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentUser: User?
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let keychain = Keychain()
    
    init() {
        checkSession()
    }
    
    func checkSession() {
        isLoading = true
        if let token = keychain.get(key: "auth_token") {
            // Validate token with backend
            validateToken(token)
        } else {
            isAuthenticated = false
            isLoading = false
        }
    }
    
    func signIn(email: String, password: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            // In production, this would call the Supabase auth API
            let response = try await APIClient.shared.signIn(email: email, password: password)
            
            if let token = response.accessToken {
                keychain.set(key: "auth_token", value: token)
                currentUser = response.user
                isAuthenticated = true
            }
        } catch {
            errorMessage = error.localizedDescription
            isAuthenticated = false
        }
        
        isLoading = false
    }
    
    func signUp(email: String, password: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            // In production, this would call the Supabase auth API
            let response = try await APIClient.shared.signUp(email: email, password: password)
            
            if let token = response.accessToken {
                keychain.set(key: "auth_token", value: token)
                currentUser = response.user
                isAuthenticated = true
            }
        } catch {
            errorMessage = error.localizedDescription
            isAuthenticated = false
        }
        
        isLoading = false
    }
    
    func signOut() {
        keychain.delete(key: "auth_token")
        currentUser = nil
        isAuthenticated = false
    }
    
    func authenticateWithBiometrics() async -> Bool {
        let context = LAContext()
        var error: NSError?
        
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            errorMessage = "Biometric authentication not available"
            return false
        }
        
        do {
            let success = try await context.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: "Authenticate to access CALQULUS RMS"
            )
            
            if success {
                // Check if user has saved credentials
                if let email = keychain.get(key: "user_email"),
                   let password = keychain.get(key: "user_password") {
                    await signIn(email: email, password: password)
                    return true
                }
            }
            
            return false
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
    
    func saveCredentials(email: String, password: String) {
        keychain.set(key: "user_email", value: email)
        keychain.set(key: "user_password", value: password)
    }
    
    private func validateToken(_ token: String) {
        // In production, this would validate the token with the backend
        // For now, we'll simulate it
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.isAuthenticated = true
            self.isLoading = false
        }
    }
}

struct User: Codable {
    let id: String
    let email: String
    let role: String
    let name: String?
}

// Keychain wrapper
class Keychain {
    func set(key: String, value: String) {
        let data = value.data(using: .utf8)!
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]
        
        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }
    
    func get(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess, let data = result as? Data {
            return String(data: data, encoding: .utf8)
        }
        
        return nil
    }
    
    func delete(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key
        ]
        
        SecItemDelete(query as CFDictionary)
    }
}

File: CALQULUS-PMS/src/mobile/google/AuthManager.kt
package com.calqulusrms.app

import android.content.Context
import android.content.SharedPreferences
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.IvParameterSpec

class AuthManager(private val context: Context) {
    private val sharedPrefs: SharedPreferences = context.getSharedPreferences("CALQULUS RMSAuth", Context.MODE_PRIVATE)
    private val keyStore = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
    
    var isAuthenticated: Boolean = false
        private set
    
    var currentUser: User? = null
        private set
    
    init {
        checkSession()
    }
    
    fun checkSession() {
        val token = sharedPrefs.getString("auth_token", null)
        if (token != null) {
            // Validate token with backend
            validateToken(token)
        } else {
            isAuthenticated = false
        }
    }
    
    suspend fun signIn(email: String, password: String): Result<User> {
        return try {
            val response = withContext(Dispatchers.IO) {
                APIClient.signIn(email, password)
            }
            
            if (response.accessToken != null) {
                sharedPrefs.edit()
                    .putString("auth_token", response.accessToken)
                    .apply()
                
                currentUser = response.user
                isAuthenticated = true
                Result.success(response.user)
            } else {
                Result.failure(Exception("Authentication failed"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun signUp(email: String, password: String): Result<User> {
        return try {
            val response = withContext(Dispatchers.IO) {
                APIClient.signUp(email, password)
            }
            
            if (response.accessToken != null) {
                sharedPrefs.edit()
                    .putString("auth_token", response.accessToken)
                    .apply()
                
                currentUser = response.user
                isAuthenticated = true
                Result.success(response.user)
            } else {
                Result.failure(Exception("Registration failed"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    fun signOut() {
        sharedPrefs.edit()
            .remove("auth_token")
            .remove("user_email")
            .remove("user_password")
            .apply()
        
        currentUser = null
        isAuthenticated = false
    }
    
    fun authenticateWithBiometrics(
        activity: FragmentActivity,
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        val biometricManager = BiometricManager.from(context)
        
        when (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)) {
            BiometricManager.BIOMETRIC_SUCCESS -> {
                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle("CALQULUS RMS Authentication")
                    .setSubtitle("Authenticate to access CALQULUS RMS")
                    .setNegativeButtonText("Cancel")
                    .build()
                
                val biometricPrompt = BiometricPrompt(
                    activity,
                    ContextCompat.getMainExecutor(context),
                    object : BiometricPrompt.AuthenticationCallback() {
                        override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                            super.onAuthenticationSucceeded(result)
                            
                            // Check if user has saved credentials
                            val email = sharedPrefs.getString("user_email", null)
                            val password = sharedPrefs.getString("user_password", null)
                            
                            if (email != null && password != null) {
                                activity.lifecycleScope.launch {
                                    val result = signIn(email, password)
                                    if (result.isSuccess) {
                                        onSuccess()
                                    } else {
                                        onError(result.exceptionOrNull()?.message ?: "Authentication failed")
                                    }
                                }
                            } else {
                                onError("No saved credentials found")
                            }
                        }
                        
                        override fun onAuthenticationFailed() {
                            super.onAuthenticationFailed()
                            onError("Authentication failed")
                        }
                        
                        override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                            super.onAuthenticationError(errorCode, errString)
                            onError(errString.toString())
                        }
                    }
                )
                
                biometricPrompt.authenticate(promptInfo)
            }
            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE,
            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> {
                onError("Biometric hardware not available")
            }
            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> {
                onError("No biometric credentials enrolled")
            }
            BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED -> {
                onError("Security update required")
            }
            BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED -> {
                onError("Biometric authentication not supported")
            }
            BiometricManager.BIOMETRIC_STATUS_UNKNOWN -> {
                onError("Unknown biometric status")
            }
        }
    }
    
    fun saveCredentials(email: String, password: String) {
        sharedPrefs.edit()
            .putString("user_email", email)
            .putString("user_password", password)
            .apply()
    }
    
    private fun validateToken(token: String) {
        // In production, this would validate the token with the backend
        // For now, we'll simulate it
        isAuthenticated = true
    }
}

data class User(
    val id: String,
    val email: String,
    val role: String,
    val name: String?
)

class AuthViewModel : ViewModel() {
    private lateinit var authManager: AuthManager
    
    val isAuthenticated: Boolean
        get() = authManager.isAuthenticated
    
    val currentUser: User?
        get() = authManager.currentUser
    
    fun init(context: Context) {
        authManager = AuthManager(context)
    }
    
    fun signOut() {
        authManager.signOut()
    }
}
